### PR TITLE
Implement Image class

### PR DESCRIPTION
**Pull request by tomczylok**
#26

### CODE REVIEW

**Comment by dkrawiec** [comment]
Extract color abstraction to another class.

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 1]
```diff
@@ -1,5 +1,103 @@
...
+        this.width = width;
+        this.height = height;
+        data = new int[width * height * 4];
+        setVisible();
+    }
+
+    private void setVisible() {
+        for(int i=0; i<width; i++) {
+            for(int j=0; j<height; j++) {
+                setAlpha(i, j, 255);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: magic number

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 2]
```diff
@@ -1,5 +1,103 @@
...
+
+    public Image() {
+        width = height = 0;
+        data = null;
+    }
+
+    public Image(final int width, final int height) {
+        this.width = width;
+        this.height = height;
+        data = new int[width * height * 4];
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: magic number

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 3]
```diff
@@ -1,5 +1,103 @@
...
+        data[getPixelBeginning(x, y) + 2] = Math.max(0, Math.min(value, 255));
+    }
+
+    public void setAlpha(final int x, final int y, final int value) {
+        data[getPixelBeginning(x, y) + 3] = Math.max(0, Math.min(value, 255));
+    }
+
+    private void checkCoordinates(final int x, final int y) {
+        if( ( x < 0 || x >= width ) || (y < 0 || y >= height) ) {
+            throw new RuntimeException("Wrong image pixel coordinates: " + x + " " + y);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: too generic exception, maybe IllegalArgumentException?

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 4]
```diff
@@ -1,5 +1,103 @@
...
+        return height;
+    }
+
+    public int[] getRGBA(final int x, final int y) {
+        checkCoordinates(x, y);
+        return new int[]{getRed(x, y), getGreen(x, y), getBlue(x, y), getAlpha(x, y)};
+    }
+
+    public void setRGBA(final int x, final int y, final int r, final int g, final int b, final int a) {
+        checkCoordinates(x, y);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: name method - check for what? maybe validateCoordinates?

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 5]
```diff
@@ -1,5 +1,103 @@
...
+        setGreen(x, y, g);
+        setBlue(x, y, b);
+    }
+
+    public int getRed(final int x, final int y) {
+        return data[getPixelBeginning(x, y)];
+    }
+
+    public int getGreen(final int x, final int y) {
+        return data[getPixelBeginning(x, y) + 1];
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: magic number

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 6]
```diff
@@ -1,5 +1,103 @@
...
+    }
+
+    public int getBlue(final int x, final int y) {
+        return data[getPixelBeginning(x, y) + 2];
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: magic number

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 7]
```diff
@@ -1,5 +1,103 @@
...
+    }
+
+    public int getAlpha(final int x, final int y) {
+        return data[getPixelBeginning(x, y) + 3];
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: magic number

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java (hunk 4 above)
Comment (dkrawiec) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: The same problem. isWithinDimmensions

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 8]
```diff
@@ -1,5 +1,103 @@
...
+    }
+
+    public void setRed(final int x, final int y, final int value) {
+        data[getPixelBeginning(x, y)] = Math.max(0, Math.min(value, 255));
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: DRY - `data[getPixelBeginning(x, y)] = Math.max(0, Math.min(value, 255));` - you only change magic number 4 times

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 9]
```diff
@@ -1,5 +1,103 @@
...
+    }
+
+    private void checkCoordinates(final int x, final int y) {
+        if( ( x < 0 || x >= width ) || (y < 0 || y >= height) ) {
+            throw new RuntimeException("Wrong image pixel coordinates: " + x + " " + y);
+        }
+    }
+
+    private int getPixelBeginning(final int x, final int y) {
+        return width * y * 4 + x * 4;
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: magic numbers

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 10]
```diff
@@ -0,0 +1,55 @@
...
+import pl.edu.uj.kimage.Image;
+import static org.assertj.core.api.Assertions.assertThat;
+
+public class ImageContainerTest {
+
+    @Test
+    public void createSampleImageObject() {
+        Image image = new Image(512, 256);
+        assertThat(image.getWidth()).isEqualTo(512);
+        assertThat(image.getHeight()).isEqualTo(256);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: test name should tell what you want to test - maybe accessorsReturnsValidValues ?

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 11]
```diff
@@ -0,0 +1,55 @@
...
+    }
+
+    @Test
+    public void oneChannelOperation() {
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: 1) The name should include which the scenario is testing whether positive or for example exception throwing. 
2) You didn't test edge cases

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+        Image image = new Image(256, 256);
+        int red = 128;
+        image.setRed(5, 5, red);
+        assertThat(image.getRed(5, 5)).isEqualTo(red);
+    }
+
+    @Test
+    public void multiChannelOperation() {
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: the same commet as above

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 13]
```diff
@@ -0,0 +1,55 @@
...
+
+        int x = 0, y = 0;
+        int red = 0, green = -5, blue = 300;
+
+        image.setRGB(0, 0, red, green, blue);
+
+        int[] colors = image.getRGB(x, y);
+
+        assertThat(colors[0]).isEqualTo(0);
+        assertThat(colors[1]).isEqualTo(0);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: could you confirm that this is correct behaviour?

Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: why this branch isn't from master?

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java (hunk 4 above)
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: :+1:

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java (hunk 13 above)
Comment (tomczylok) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: I reckon that we should clamp values extending borders (min = 0, max = 255). For example when we add two images, values many times exceed value 255 which usually means that pixel should be pure white (255).

Branch from feature/16-StepMechanism, cause it already had empty Image class prepared, so why not.

Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java]: it could be tricky when you will need to rebase onto master ;)

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java (hunk 3 above)
Comment (tomczylok) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: IndexOutOfBoundsException?

Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java]: Not so bad, but that will suggest usage of array

**Comment by dkrawiec** [comment]
I think it would be nice to provide builder for Image class so that you don't have to provide array of Color on your own.

**Comment by dkrawiec** [comment]
I guess we need another implementation of a Color class. Which is GrayscaleColor. 

Actually we could extend the approach with static fields to a flyweight design pattern, since this class is immutable, and actually a value object.

**Comment by kmisztal** [comment]
ok

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Color.java [hunk 14]
```diff
@@ -0,0 +1,39 @@
+package pl.edu.uj.kimage;
+
+public final class Color {
+    private final int red, green, blue, alpha;
+
+    public static final int minColorValue = 0;
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Color.java:6]: Did you use formatter? static should be above non-static and public above private

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 15]
```diff
@@ -1,29 +1,23 @@
...
 
     public Image() {
         width = height = 0;
         data = null;
     }
 
-    public Image(final int width, final int height) {
+    public Image(final int width, final int height, final Color[] data) {
+        if(data.length != width * height) {
+            throw new IllegalArgumentException("Size of image does not match with input color data size.");
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java:15]: `Add message to each exception with explanation of cause with all additional data. For example when the file was not found add path to this file in exception message.` from `CONTRIBUTING.md`

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 16]
```diff
@@ -34,70 +28,45 @@ public int getHeight() {
...
-    }
-
-    public void setRGBA(final int x, final int y, final int r, final int g, final int b, final int a) {
-        checkCoordinates(x, y);
-        setRed(x, y, r);
-        setGreen(x, y, g);
-        setBlue(x, y, b);
-        setAlpha(x, y, a);
+    public Color[] getData() {
+        return data;
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java:32]: `Arrays.copyOf()`?

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 17]
```diff
@@ -34,70 +28,45 @@ public int getHeight() {
...
     }
 
-    public int[] getRGB(final int x, final int y) {
-        checkCoordinates(x, y);
-        return new int[]{getRed(x, y), getGreen(x, y), getBlue(x, y)};
+    public Color getRGB(final int x, final int y) {
+        isWithinDimensions(x, y);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java:36]: looks like getter, but ok

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 18]
```diff
@@ -1,55 +1,58 @@
...
 public class ImageContainerTest {
 
     @Test
-    public void createSampleImageObject() {
-        Image image = new Image(512, 256);
-        assertThat(image.getWidth()).isEqualTo(512);
-        assertThat(image.getHeight()).isEqualTo(256);
+    public void sizeAccessorsReturnValidValues() {
+        int width = 2, height = 2;
+        Color[] colors = new Color[]{new Color(255, 0, 0), new Color(0, 255, 0),
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java:13]: magic numbers

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 19]
```diff
@@ -1,55 +1,58 @@
...
-        assertThat(image.getRed(5, 5)).isEqualTo(red);
+    public void channelColorAccessorsReturnValidValues() {
+        int width = 1, height = 1;
+        int redChannel = 255, greenChannel = 0, blueChannel = 0;
+        Color[] colors = new Color[]{new Color(redChannel, greenChannel, blueChannel)};
+
+        Image image = new Image(width, height, colors);
+        assertThat(image.getRed(0, 0)).isEqualTo(redChannel);
+        assertThat(image.getGreen(0, 0)).isEqualTo(greenChannel);
+        assertThat(image.getBlue(0, 0)).isEqualTo(blueChannel);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java:30]: many magic numbers

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 20]
```diff
@@ -1,55 +1,58 @@
...
-        assertThat(colors[2]).isEqualTo(blue);
+    public void channelColorAccessorsReturnClampedValues() {
+        int width = 1, height = 1;
+        int redChannel = 0, greenChannel = -25, blueChannel = 280;
+        Color[] colors = new Color[]{new Color(redChannel, greenChannel, blueChannel)};
+
+        Image image = new Image(width, height, colors);
+        assertThat(image.getRed(0, 0)).isEqualTo(0);
+        assertThat(image.getGreen(0, 0)).isEqualTo(0);
+        assertThat(image.getBlue(0, 0)).isEqualTo(255);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java:42]: many magic numbers

File: imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java [hunk 21]
```diff
@@ -1,55 +1,58 @@
...
+        int width = 2, height = 2;
+        Color[] colors = new Color[]{new Color(255, 0, 0), new Color(0, 255, 0),
+                                     new Color(0, 0, 255), new Color(255, 0, 255)};
+
+        Image image = new Image(width, height, colors);
+        Color rgba = image.getRGBA(1, 1);
+        assertThat(rgba.getRed()).isEqualTo(255);
+        assertThat(rgba.getGreen()).isEqualTo(0);
+        assertThat(rgba.getBlue()).isEqualTo(255);
+        assertThat(rgba.getAlpha()).isEqualTo(255);
```
Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java:56]: many magic numbers

Comment (annaskulimowska) [inline imageprocessor/imageprocessor-core/src/test/java/pl/edu/uj/kimage/processingFlow/ImageContainerTest.java:56]: where is the test for
`throw new IllegalArgumentException("Size of image does not match with input color data size.");` and for `throw new IllegalArgumentException("Wrong image pixel coordinates: " + x + " " + y);`?

Testing only positive scenario does not make sense.

File: imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java [hunk 22]
```diff
@@ -1,5 +1,72 @@
...
+
+    public int getAlpha(final int x, final int y) {
+        return data[getPixelIndex(x, y)].getAlpha();
+    }
+
+    private int getPixelIndex(final int x, final int y) {
+        return y * width + x;
+    }
+
+    private void isWithinDimensions(final int x, final int y) {
```
Comment (dkrawiec) [inline imageprocessor/imageprocessor-core/src/main/java/pl/edu/uj/kimage/Image.java:67]: please make this method public

**Comment by dkrawiec** [comment]
@tomczylok please fix mentioned issues and open another pull request, this time to master.
